src/blog/pointer-compression.md

```diff
@@ -0,0 +1,537 @@
+
+There are two operations that are crucial for Pointer Compression: Loads and Stores, since they connect the V8 heap with the rest of the pipeline. If we were to decompress every time we loaded a compressed value from the heap, and compress it before we store it, then the pipeline could just keep working as it otherwise did in full-pointer mode. Thus we added new explicit value operations in the node graph - Decompress and Compress.
+
+There are cases where the decompression is not actually necessary. For example, if a compressed value is loaded from somewhere only to be then stored to a new location.
+
+In order to optimize unnecessary operations, we implemented a new “Decompression Elimination” phase in TurboFan. Its job is to eliminate decompressions directly followed by compressions. Since these nodes might not be directly next to each other it also tries to propagate decompressions through the graph, with the hope of encountering a compress down the line and eliminating them both. This gave us a 2% improvement of Octane’ score.
+
+#### Bump (3), +2%
+
+While we were looking at the generated code, we noticed that the decompression of a value that had just been loaded produced code that was a bit too verbose:
+
+```asm
+movl rax, <mem>   // load
+movlsxlq rax, rax // sign extend
+```
+
+Once we fixed that to sign extend the value loaded from memory directly:
+
+```asm
+movlsxlq rax, <mem>
+```
+
+so got yet another 2% improvement.
+
+##### Bump (4), +11%
+
+TurboFan optimization phases work by using pattern matching on the graph: once a sub-graph matches a certain pattern it is replaced with a semantically equivalent (but better) sub-graph or instruction.
+
+Unsuccessful attempts to find a match are not an explicit failure. The presence of explicit Decompress/Compress operations in the graph caused previously successful pattern matching attempts to no longer succeed, resulting in optimizations silently failing.
+
+One example of a “broken” optimization was [allocation preternuring](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43823.pdf). Once we updated the pattern matching to be aware of the new compression/decompression nodes we got another 11% improvement.
+
+### Further improvements
+
+<figure>
+  <img src="/_img/pointer-compression/perf-octane-2.svg" width="859" height="178" alt="Second round of Octane's improvements" loading="lazy">
+  <figcaption>Second round of Octane's improvements</figcaption>
+</figure>
+
+#### Bump (5), +0.5%
+
+While implementing the Decompression Elimination in TurboFan we learned a lot. The explicit Decompression/Compression node approach had the following properties:
+
+Pros:
+    - Explicitness of such operations allowed us to optimize unnecessary decompressions by doing canonical pattern matching of sub-graphs.
```

Comment:
These lists don't render as actual lists, you might want to add a newline before the first item.